<a href="https://colab.research.google.com/github/ksbharaj/COMP2231/blob/main/Week_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib.pyplot as plt

import plotly.graph_objects as go

import plotly.express as px
import datetime as dt

In [ ]:
## Pandas Setting for easier viewability
pd.set_option('display.max_columns', None)
pd.options.mode.chained_assignment = None

In [ ]:
## Import df_users dataset
df_users = pd.read_csv('/content/drive/MyDrive/DataCAN/df_users.csv')

In [ ]:
## Change the data types after importing for future convenience
df_users.signup_date = pd.to_datetime(df_users.signup_date)
df_users.trial_date = pd.to_datetime(df_users.trial_date)
df_users.subscr_3m_date = pd.to_datetime(df_users.subscr_3m_date)
df_users.subscr_12m_date = pd.to_datetime(df_users.subscr_12m_date)
df_users.signup_date_new = pd.to_datetime(df_users.signup_date_new)

In [ ]:
## Import df_users dataset
df_campaign_cost = pd.read_csv('/content/drive/MyDrive/DataCAN/df_campaign_cost.csv')
df_campaigns = pd.read_csv('/content/drive/MyDrive/DataCAN/df_campaigns.csv')
df_conversion = pd.read_csv('/content/drive/MyDrive/DataCAN/df_conversion.csv')
df_visit = pd.read_csv('/content/drive/MyDrive/DataCAN/df_visit.csv')

In [ ]:
df_campaign_cost.year_month_day = pd.to_datetime(df_campaign_cost.year_month_day)

In [ ]:
df_campaign_cost.sort_values(by="clicks", ascending=False)

Unnamed: 0 year_month_day  campaign_id  impressions  clicks    cost
7846        7846     2022-09-11       233290       2855.0      41  188.56
7586        7586     2022-08-16       233290       5779.0      38  174.77
8006        8006     2022-09-27       233290       6535.0      37  220.55
7506        7506     2022-08-08       233290       6336.0      37  189.17
7926        7926     2022-09-19       233290       3769.0      36  201.65
...          ...            ...          ...          ...     ...     ...
2204        2204     2020-12-27       158347        199.0       1    3.62
2202        2202     2020-12-26       727345        120.0       1    3.30
2200        2200     2020-12-26       710412         63.0       1    1.80
2198        2198     2020-12-26       233290        175.0       1    5.41
0              0     2019-01-01       845442        145.0       1    3.17

[8660 rows x 6 columns]

In [ ]:
df_campaign_cost.campaign_id.unique()

array([845442, 660013, 158347, 710412, 233290, 727345, 654321, 989221,
       443618, 560121])

In [ ]:
df_campaigns

Unnamed: 0  campaign_id                        campaign_name  \
0           0       727345           trial_recruitment_campaign   
1           1       654321             fact_or_fiction_campaign   
2           2       989221                     try_out_campaign   
3           3       443618             become_a_member_campaign   
4           4       710412          limited_time_offer_campaign   
5           5       845442     gain_from_using_product_campaign   
6           6       158347                  click_fast_campaign   
7           7       560121  colleagues_will_be_jealous_campaign   
8           8       660013       nine_reasons_to_click_campaign   
9           9       233290             free_week_bonus_campaign   

  campaign_status  
0          active  
1          active  
2          active  
3          active  
4          active  
5          active  
6          active  
7          active  
8          active  
9          active

In [ ]:
len(set(df_conversion.user_id).intersection(set(df_users.user_id)))

43898

In [ ]:
df_campaign_cost['year_month'] = df_campaign_cost['year_month_day'].dt.to_period('M')

In [ ]:
df_campaign_cost['year_month_str'] = df_campaign_cost['year_month'].astype(str)

In [ ]:
df_campaign_cost = pd.merge(df_campaign_cost, df_campaigns, on='campaign_id')

In [ ]:
grouped_df = df_campaign_cost.groupby(['campaign_name', 'year_month_str']).sum().reset_index()

<ipython-input-15-19e463ef2016>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  grouped_df = df_campaign_cost.groupby(['campaign_name', 'year_month_str']).sum().reset_index()


In [ ]:
fig1 = px.line(grouped_df, x='year_month_str', y='cost', color='campaign_name', title='Total Cost by Campaign Name and Month & Year')
fig1.show()

In [ ]:
# Create the cumulative plot
grouped_df['cumulative_cost'] = grouped_df.groupby('campaign_name')['cost'].cumsum()
fig2 = px.line(grouped_df, x='year_month_str', y='cumulative_cost', color='campaign_name', title='Cumulative Cost by Campaign Name and Month & Year')
fig2.show()

In [ ]:
fig3 = px.line(grouped_df, x='year_month_str', y='impressions', color='campaign_name', title='Total Cost by Campaign Name and Month & Year')
fig3.show()

In [ ]:
grouped_df['cumulative_impressions'] = grouped_df.groupby('campaign_name')['impressions'].cumsum()
fig4 = px.line(grouped_df, x='year_month_str', y='cumulative_impressions', color='campaign_name', title='Cumulative Impressions by Campaign Name and Month & Year')
fig4.show()

In [ ]:
fig5 = px.line(grouped_df, x='year_month_str', y='clicks', color='campaign_name', title='Total Clicks by Campaign Name and Month & Year')
fig5.show()

In [ ]:
grouped_df['cumulative_clicks'] = grouped_df.groupby('campaign_name')['clicks'].cumsum()
fig6 = px.line(grouped_df, x='year_month_str', y='cumulative_clicks', color='campaign_name', title='Cumulative Clicks by Campaign Name and Month & Year')
fig6.show()

"Free_week_bonus_campagin" is clearly more expensive that the others. Their usage was not equal acorss time. There is a notable increase in usage starting Jan 2022. THis can be said for all campaigns, peaking at Sep 2022.

## Question 2: Based on the data, which campaign has the best cost per click?

In [ ]:
df_campaign_cost['cpc'] = df_campaign_cost['cost'] / df_campaign_cost['clicks']

In [ ]:
grouped_df = df_campaign_cost.groupby(['campaign_name', 'year_month_str']).mean().reset_index()

<ipython-input-23-4748a1223177>:1: FutureWarning:

The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.



In [ ]:
fig_cpc = px.line(grouped_df, x='year_month_str', y='cpc', color='campaign_name', title='Average Cost-Per-Click by Campaign Name and Month & Year')
fig_cpc.show()

"Free_week_bonus_campagin" and "Nine_reasons_to_click_campaign" are have the highest cost per click ratio

## Question 3: Identify email(s) with issues, better open rate, better click rate and faster CTA

In [ ]:
df_sent = pd.read_csv('/content/drive/MyDrive/DataCAN/df_sent.csv', index_col=0)
df_delivered = pd.read_csv('/content/drive/MyDrive/DataCAN/df_delivered.csv', index_col=0)
df_open = pd.read_csv('/content/drive/MyDrive/DataCAN/df_open.csv', index_col=0)
df_clicked = pd.read_csv('/content/drive/MyDrive/DataCAN/df_clicked.csv', index_col=0)
df_spam = pd.read_csv('/content/drive/MyDrive/DataCAN/df_spam.csv', index_col=0)
df_bounced = pd.read_csv('/content/drive/MyDrive/DataCAN/df_bounced.csv', index_col=0)

In [ ]:
email_counts = df_sent['emails'].value_counts().reset_index()
email_counts.columns = ['emails', 'count']
fig = px.bar(email_counts, x='emails', y='count', title='Number of Rows for Each Type of Email')
fig.show()

In [ ]:
df_delivered

user_id                             emails      delivered_email
0      832804001                   Welcome New User  2021-02-21 01:37:27
1      287204540                   Welcome New User  2020-05-27 00:57:14
2      959081297                   Welcome New User  2021-04-08 14:06:17
3      691397025                   Welcome New User  2019-07-19 13:55:51
4      994470486                   Welcome New User  2019-04-11 04:35:56
...          ...                                ...                  ...
37935  489590675  How You Can Optimize Your Writing  2020-08-11 17:00:32
37936  562131477  How You Can Optimize Your Writing  2021-08-13 08:42:15
37937  943865726  How You Can Optimize Your Writing  2021-06-07 15:09:17
37938  895380653  How You Can Optimize Your Writing  2020-07-07 22:03:38
37939  490237820  How You Can Optimize Your Writing  2020-09-08 11:15:53

[37940 rows x 3 columns]

In [ ]:
merged_df = pd.merge(df_sent, df_delivered, on=['user_id', 'emails'], how='outer')

In [ ]:
merged_df = pd.merge(merged_df, df_open, on=['user_id', 'emails'], how='outer')

In [ ]:
merged_df = pd.merge(merged_df, df_clicked, on=['user_id', 'emails'], how='outer')
merged_df = pd.merge(merged_df, df_spam, on=['user_id', 'emails'], how='outer')
merged_df = pd.merge(merged_df, df_bounced, on=['user_id', 'emails'], how='outer')

In [ ]:
merged_df

user_id                             emails           sent_email  \
0      832804001                   Welcome New User  2021-02-21 01:37:27   
1      287204540                   Welcome New User  2020-05-27 00:57:14   
2      959081297                   Welcome New User  2021-04-08 14:06:17   
3      691397025                   Welcome New User  2019-07-19 13:55:51   
4      994470486                   Welcome New User  2019-04-11 04:35:56   
...          ...                                ...                  ...   
42740  562131477  How You Can Optimize Your Writing  2021-08-13 08:42:15   
42741  943865726  How You Can Optimize Your Writing  2021-06-07 15:09:17   
42742  555300490  How You Can Optimize Your Writing  2021-08-17 00:44:18   
42743  895380653  How You Can Optimize Your Writing  2020-07-07 22:03:38   
42744  490237820  How You Can Optimize Your Writing  2020-09-08 11:15:53   

           delivered_email                  open_email clicked_email  \
0      2021-02-21 01:37:27                         NaN           NaN   
1      2020-05-27 00:57:14                         NaN           NaN   
2      2021-04-08 14:06:17                         NaN           NaN   
3      2019-07-19 13:55:51  2019-07-24 05:55:21.193460           NaN   
4      2019-04-11 04:35:56                         NaN           NaN   
...                    ...                         ...           ...   
42740  2021-08-13 08:42:15                         NaN           NaN   
42741  2021-06-07 15:09:17                         NaN           NaN   
42742                  NaN                         NaN           NaN   
42743  2020-07-07 22:03:38                         NaN           NaN   
42744  2020-09-08 11:15:53                         NaN           NaN   

                spam_email        bounced_email  
0                      NaN                  NaN  
1                      NaN                  NaN  
2                      NaN                  NaN  
3                      NaN                  NaN  
4                      NaN                  NaN  
...                    ...                  ...  
42740                  NaN                  NaN  
42741                  NaN                  NaN  
42742                  NaN  2021-08-17 00:44:18  
42743  2020-07-07 22:03:38                  NaN  
42744                  NaN                  NaN  

[42745 rows x 8 columns]

In [ ]:
sum(merged_df.sent_email.notna())

42745

In [ ]:
## Delivered only
sum(merged_df.delivered_email.notna() & merged_df.open_email.isna() & merged_df.clicked_email.isna() & merged_df.spam_email.isna()
    & merged_df.bounced_email.isna())

30177

In [ ]:
## Bounced only
sum(merged_df.delivered_email.isna() & merged_df.open_email.isna() & merged_df.clicked_email.isna() & merged_df.spam_email.isna()
    & merged_df.bounced_email.notna())

4805

In [ ]:
## Delivered and Opened only
sum(merged_df.delivered_email.notna() & merged_df.open_email.notna() & merged_df.clicked_email.isna() & merged_df.spam_email.isna()
    & merged_df.bounced_email.isna())

5905

In [ ]:
## Delivered and Spam only
sum(merged_df.delivered_email.notna() & merged_df.open_email.isna() & merged_df.clicked_email.isna() & merged_df.spam_email.notna()
    & merged_df.bounced_email.isna())

1416

In [ ]:
## Delivered, opened and Clicked only
sum(merged_df.delivered_email.notna() & merged_df.open_email.notna() & merged_df.clicked_email.notna() & merged_df.spam_email.isna()
    & merged_df.bounced_email.isna())

442

SO each email has any one of the 5 possible outcomes:
- Delivered Only
- Bounced Only
- Delivered and Opened Only
- Delivered and Spam
- Delivered, Opened and Clicked

In [ ]:
email_type_dict = {}

In [ ]:
for index, row in merged_df.iterrows():
  email_type = row['emails']
  if email_type not in email_type_dict:
    email_type_dict[email_type] = {'Delivered Only': 0, 'Bounced Only': 0, 'Delievered and Opened Only': 0,
                                   'Delivered and Spam': 0, 'Delivered, Opened and Clicked': 0}
  if pd.notna(row['bounced_email']):
    email_type_dict[email_type]['Bounced Only'] += 1
  if pd.notna(row['delivered_email']) & pd.isna(row['open_email']) & pd.isna(row['clicked_email']) & pd.isna(row['spam_email']) & \
     pd.isna(row['bounced_email']):
     email_type_dict[email_type]['Delivered Only'] += 1
  if pd.notna(row['delivered_email']) & pd.notna(row['open_email']) & pd.isna(row['clicked_email']) & pd.isna(row['spam_email']) & \
     pd.isna(row['bounced_email']):
     email_type_dict[email_type]['Delievered and Opened Only'] += 1
  if pd.notna(row['delivered_email']) & pd.isna(row['open_email']) & pd.isna(row['clicked_email']) & pd.notna(row['spam_email']) & \
     pd.isna(row['bounced_email']):
     email_type_dict[email_type]['Delivered and Spam'] += 1
  if pd.notna(row['delivered_email']) & pd.notna(row['open_email']) & pd.notna(row['clicked_email']) & pd.isna(row['spam_email']) & \
     pd.isna(row['bounced_email']):
     email_type_dict[email_type]['Delivered, Opened and Clicked'] += 1

In [ ]:
email_status_df = pd.DataFrame(email_type_dict).transpose().reset_index()
email_status_df.rename(columns={'index': 'Email Type'}, inplace=True)

In [ ]:
fig = px.bar(email_status_df, x='Email Type', y=['Delivered Only', 'Bounced Only', 'Delievered and Opened Only', 'Delivered and Spam',
                                                 'Delivered, Opened and Clicked'],
             title='Email Status by Email Type',
             labels={'value': 'Number of Emails'},
             height=600)
fig.show()

In terms of issues, "how You Can Optimize Your Writing" had the most bounced emails

In [ ]:
email_status_df

Email Type  Delivered Only  Bounced Only  \
0                   Welcome New User            8474           220   
1     Error Free Messages Benefits 1            4303           137   
2     Error Free Messages Benefits 2            3906           214   
3               Long Weekend Special            7356           353   
4  How You Can Optimize Your Writing            6138          3881   

   Delievered and Opened Only  Delivered and Spam  \
0                        1596                 225   
1                         631                 115   
2                        1142                 166   
3                        1329                 661   
4                        1207                 249   

   Delivered, Opened and Clicked  
0                            175  
1                             48  
2                             28  
3                             64  
4                            127

In [ ]:
import copy

df_proportions = copy.deepcopy(email_status_df)

In [ ]:
df_proportions['Total'] = df_proportions.iloc[:, 1:].sum(axis=1)
y_prop_cols = [col + ' (Y Prop)' for col in df_proportions.columns[1:-1]]
df_proportions[y_prop_cols] = df_proportions.iloc[:, 1:-1].div(df_proportions['Total'], axis=0)
column_sums = df_proportions.iloc[:, 1:-1].sum()
x_prop = column_sums / column_sums.sum()
x_prop = x_prop.rename(lambda x: x + ' (X Prop)')
df_final = df_proportions.copy()
df_final.loc['Total'] = column_sums.append(x_prop)
df_final

<ipython-input-44-3b72096833ae>:8: FutureWarning:

The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.



Email Type  Delivered Only  Bounced Only  \
0                       Welcome New User          8474.0         220.0   
1         Error Free Messages Benefits 1          4303.0         137.0   
2         Error Free Messages Benefits 2          3906.0         214.0   
3                   Long Weekend Special          7356.0         353.0   
4      How You Can Optimize Your Writing          6138.0        3881.0   
Total                                NaN         30177.0        4805.0   

       Delievered and Opened Only  Delivered and Spam  \
0                          1596.0               225.0   
1                           631.0               115.0   
2                          1142.0               166.0   
3                          1329.0               661.0   
4                          1207.0               249.0   
Total                      5905.0              1416.0   

       Delivered, Opened and Clicked    Total  Delivered Only (Y Prop)  \
0                              175.0  10690.0                 0.792703   
1                               48.0   5234.0                 0.822125   
2                               28.0   5456.0                 0.715909   
3                               64.0   9763.0                 0.753457   
4                              127.0  11602.0                 0.529047   
Total                          442.0  42745.0                 3.613241   

       Bounced Only (Y Prop)  Delievered and Opened Only (Y Prop)  \
0                   0.020580                             0.149298   
1                   0.026175                             0.120558   
2                   0.039223                             0.209311   
3                   0.036157                             0.136126   
4                   0.334511                             0.104034   
Total               0.456646                             0.719327   

       Delivered and Spam (Y Prop)  Delivered, Opened and Clicked (Y Prop)  
0                         0.021048                                0.016370  
1                         0.021972                                0.009171  
2                         0.030425                                0.005132  
3                         0.067705                                0.006555  
4                         0.021462                                0.010946  
Total                     0.162611                                     NaN

In [ ]:
df_final["Open Rate"] = df_final["Delievered and Opened Only (Y Prop)"] + df_final["Delivered, Opened and Clicked (Y Prop)"]

In [ ]:
df_final

Email Type  Delivered Only  Bounced Only  \
0                       Welcome New User          8474.0         220.0   
1         Error Free Messages Benefits 1          4303.0         137.0   
2         Error Free Messages Benefits 2          3906.0         214.0   
3                   Long Weekend Special          7356.0         353.0   
4      How You Can Optimize Your Writing          6138.0        3881.0   
Total                                NaN         30177.0        4805.0   

       Delievered and Opened Only  Delivered and Spam  \
0                          1596.0               225.0   
1                           631.0               115.0   
2                          1142.0               166.0   
3                          1329.0               661.0   
4                          1207.0               249.0   
Total                      5905.0              1416.0   

       Delivered, Opened and Clicked    Total  Delivered Only (Y Prop)  \
0                              175.0  10690.0                 0.792703   
1                               48.0   5234.0                 0.822125   
2                               28.0   5456.0                 0.715909   
3                               64.0   9763.0                 0.753457   
4                              127.0  11602.0                 0.529047   
Total                          442.0  42745.0                 3.613241   

       Bounced Only (Y Prop)  Delievered and Opened Only (Y Prop)  \
0                   0.020580                             0.149298   
1                   0.026175                             0.120558   
2                   0.039223                             0.209311   
3                   0.036157                             0.136126   
4                   0.334511                             0.104034   
Total               0.456646                             0.719327   

       Delivered and Spam (Y Prop)  Delivered, Opened and Clicked (Y Prop)  \
0                         0.021048                                0.016370   
1                         0.021972                                0.009171   
2                         0.030425                                0.005132   
3                         0.067705                                0.006555   
4                         0.021462                                0.010946   
Total                     0.162611                                     NaN   

       Open Rate  
0       0.165669  
1       0.129729  
2       0.214443  
3       0.142682  
4       0.114980  
Total        NaN

"Error Free Messages Benefits 2" has the highest Open Rate of 21.44%

"Welcome New User" has the highest Click Rate at 1.64%



In [ ]:
clicked_df = merged_df[merged_df.clicked_email.notna()]

In [ ]:
clicked_df

user_id                             emails           sent_email  \
97     972496068                   Welcome New User  2020-10-27 07:06:28   
148    972129985                   Welcome New User  2019-01-09 03:13:58   
183    942492107                   Welcome New User  2019-02-06 00:35:06   
195    900343660                   Welcome New User  2021-08-06 23:13:24   
204    788735789                   Welcome New User  2020-07-29 18:52:10   
...          ...                                ...                  ...   
42516  535953666  How You Can Optimize Your Writing  2020-06-23 09:55:49   
42545  316660361  How You Can Optimize Your Writing  2020-11-26 21:16:14   
42568  213402413  How You Can Optimize Your Writing  2022-01-23 19:28:46   
42582  285505196  How You Can Optimize Your Writing  2021-11-15 10:23:15   
42665  669424706  How You Can Optimize Your Writing  2022-07-24 23:54:15   

           delivered_email                  open_email  \
97     2020-10-27 07:06:28  2020-10-28 01:42:05.821716   
148    2019-01-09 03:13:58  2019-01-14 01:42:11.230223   
183    2019-02-06 00:35:06  2019-02-08 22:22:14.077332   
195    2021-08-06 23:13:24  2021-08-09 06:23:09.780690   
204    2020-07-29 18:52:10  2020-07-31 04:16:41.565331   
...                    ...                         ...   
42516  2020-06-23 09:55:49  2020-06-25 00:45:49.019119   
42545  2020-11-26 21:16:14  2020-11-30 04:20:15.199652   
42568  2022-01-23 19:28:46  2022-01-28 11:58:32.588654   
42582  2021-11-15 10:23:15  2021-11-19 11:16:16.809318   
42665  2022-07-24 23:54:15  2022-07-29 12:47:53.517474   

                    clicked_email spam_email bounced_email  
97     2020-10-28 01:42:26.696669        NaN           NaN  
148    2019-01-14 01:42:25.607778        NaN           NaN  
183    2019-02-08 22:22:39.711449        NaN           NaN  
195    2021-08-09 06:23:39.722770        NaN           NaN  
204    2020-07-31 04:16:57.258438        NaN           NaN  
...                           ...        ...           ...  
42516  2020-06-25 00:46:11.796510        NaN           NaN  
42545  2020-11-30 04:20:39.074369        NaN           NaN  
42568  2022-01-28 11:58:56.911358        NaN           NaN  
42582  2021-11-19 11:16:31.786194        NaN           NaN  
42665  2022-07-29 12:48:22.349081        NaN           NaN  

[442 rows x 8 columns]

In [ ]:
clicked_df['sent_email'] = pd.to_datetime(clicked_df['sent_email'])
clicked_df['clicked_email'] = pd.to_datetime(clicked_df['clicked_email'])

In [ ]:
clicked_df['time_to_click'] = clicked_df['clicked_email'] - clicked_df['sent_email']
clicked_df[['user_id', 'emails', 'time_to_click']]

user_id                             emails          time_to_click
97     972496068                   Welcome New User 0 days 18:35:58.696669
148    972129985                   Welcome New User 4 days 22:28:27.607778
183    942492107                   Welcome New User 2 days 21:47:33.711449
195    900343660                   Welcome New User 2 days 07:10:15.722770
204    788735789                   Welcome New User 1 days 09:24:47.258438
...          ...                                ...                    ...
42516  535953666  How You Can Optimize Your Writing 1 days 14:50:22.796510
42545  316660361  How You Can Optimize Your Writing 3 days 07:04:25.074369
42568  213402413  How You Can Optimize Your Writing 4 days 16:30:10.911358
42582  285505196  How You Can Optimize Your Writing 4 days 00:53:16.786194
42665  669424706  How You Can Optimize Your Writing 4 days 12:54:07.349081

[442 rows x 3 columns]

In [ ]:
clicked_df['time_to_click_mins'] = clicked_df['time_to_click'].dt.total_seconds() / 60

In [ ]:
fig = px.histogram(clicked_df, x="time_to_click_mins", color="emails",nbins=100,
                   labels={'time_to_click_mins': 'Time to Click (minutes)', 'count': 'Count'},
                   title='Distribution of Time to Click for Different Email Types')

# Show the figure
fig.show()

In [ ]:
clicked_df

user_id                             emails          sent_email  \
97     972496068                   Welcome New User 2020-10-27 07:06:28   
148    972129985                   Welcome New User 2019-01-09 03:13:58   
183    942492107                   Welcome New User 2019-02-06 00:35:06   
195    900343660                   Welcome New User 2021-08-06 23:13:24   
204    788735789                   Welcome New User 2020-07-29 18:52:10   
...          ...                                ...                 ...   
42516  535953666  How You Can Optimize Your Writing 2020-06-23 09:55:49   
42545  316660361  How You Can Optimize Your Writing 2020-11-26 21:16:14   
42568  213402413  How You Can Optimize Your Writing 2022-01-23 19:28:46   
42582  285505196  How You Can Optimize Your Writing 2021-11-15 10:23:15   
42665  669424706  How You Can Optimize Your Writing 2022-07-24 23:54:15   

           delivered_email                  open_email  \
97     2020-10-27 07:06:28  2020-10-28 01:42:05.821716   
148    2019-01-09 03:13:58  2019-01-14 01:42:11.230223   
183    2019-02-06 00:35:06  2019-02-08 22:22:14.077332   
195    2021-08-06 23:13:24  2021-08-09 06:23:09.780690   
204    2020-07-29 18:52:10  2020-07-31 04:16:41.565331   
...                    ...                         ...   
42516  2020-06-23 09:55:49  2020-06-25 00:45:49.019119   
42545  2020-11-26 21:16:14  2020-11-30 04:20:15.199652   
42568  2022-01-23 19:28:46  2022-01-28 11:58:32.588654   
42582  2021-11-15 10:23:15  2021-11-19 11:16:16.809318   
42665  2022-07-24 23:54:15  2022-07-29 12:47:53.517474   

                   clicked_email spam_email bounced_email  \
97    2020-10-28 01:42:26.696669        NaN           NaN   
148   2019-01-14 01:42:25.607778        NaN           NaN   
183   2019-02-08 22:22:39.711449        NaN           NaN   
195   2021-08-09 06:23:39.722770        NaN           NaN   
204   2020-07-31 04:16:57.258438        NaN           NaN   
...                          ...        ...           ...   
42516 2020-06-25 00:46:11.796510        NaN           NaN   
42545 2020-11-30 04:20:39.074369        NaN           NaN   
42568 2022-01-28 11:58:56.911358        NaN           NaN   
42582 2021-11-19 11:16:31.786194        NaN           NaN   
42665 2022-07-29 12:48:22.349081        NaN           NaN   

               time_to_click  time_to_click_mins  
97    0 days 18:35:58.696669         1115.978278  
148   4 days 22:28:27.607778         7108.460130  
183   2 days 21:47:33.711449         4187.561857  
195   2 days 07:10:15.722770         3310.262046  
204   1 days 09:24:47.258438         2004.787641  
...                      ...                 ...  
42516 1 days 14:50:22.796510         2330.379942  
42545 3 days 07:04:25.074369         4744.417906  
42568 4 days 16:30:10.911358         6750.181856  
42582 4 days 00:53:16.786194         5813.279770  
42665 4 days 12:54:07.349081         6534.122485  

[442 rows x 10 columns]

In [ ]:
average_time_to_click = clicked_df.groupby('emails')['time_to_click_mins'].mean().reset_index()
average_time_to_click.rename(columns={'time_to_click_mins': 'Average Time to Click (mins)'}, inplace=True)
average_time_to_click

emails  Average Time to Click (mins)
0     Error Free Messages Benefits 1                   5466.962331
1     Error Free Messages Benefits 2                   5260.434000
2  How You Can Optimize Your Writing                   5026.420976
3               Long Weekend Special                   5078.488341
4                   Welcome New User                   5157.326752

From time the email is sent to the time it is clicked, "How You Can Optimize Your Writing" takes teh least time at just over 5000 minutes. However, there isn't alot to seperate all of them

---



## Question 4: Was the email "warm-up" process performed correctly?

In [ ]:
merged_df.sort_values(by="sent_email")

user_id                             emails           sent_email  \
21510  914162418               Long Weekend Special  2019-01-01 15:08:20   
31364  395476300  How You Can Optimize Your Writing  2019-01-01 17:02:19   
311    589260571                   Welcome New User  2019-01-02 17:55:57   
31432  596201111  How You Can Optimize Your Writing  2019-01-04 15:03:42   
21560  890117175               Long Weekend Special  2019-01-05 04:50:22   
...          ...                                ...                  ...   
15421  705295872     Error Free Messages Benefits 1  2022-12-03 12:40:28   
21058  390792121     Error Free Messages Benefits 2  2022-12-05 12:34:46   
20980  225836231     Error Free Messages Benefits 2  2022-12-05 23:17:07   
15457  492278235     Error Free Messages Benefits 1  2022-12-06 23:26:50   
21077  553355367     Error Free Messages Benefits 2  2022-12-08 08:00:07   

           delivered_email                  open_email  \
21510  2019-01-01 15:08:20                         NaN   
31364                  NaN                         NaN   
311    2019-01-02 17:55:57                         NaN   
31432  2019-01-04 15:03:42                         NaN   
21560  2019-01-05 04:50:22                         NaN   
...                    ...                         ...   
15421  2022-12-03 12:40:28                         NaN   
21058  2022-12-05 12:34:46  2022-12-06 23:59:34.584360   
20980  2022-12-05 23:17:07                         NaN   
15457  2022-12-06 23:26:50                         NaN   
21077  2022-12-08 08:00:07  2022-12-12 04:13:56.609967   

                    clicked_email           spam_email        bounced_email  
21510                         NaN                  NaN                  NaN  
31364                         NaN                  NaN  2019-01-01 17:02:19  
311                           NaN                  NaN                  NaN  
31432                         NaN  2019-01-04 15:03:42                  NaN  
21560                         NaN                  NaN                  NaN  
...                           ...                  ...                  ...  
15421                         NaN                  NaN                  NaN  
21058                         NaN                  NaN                  NaN  
20980                         NaN                  NaN                  NaN  
15457                         NaN                  NaN                  NaN  
21077  2022-12-12 04:14:20.068286                  NaN                  NaN  

[42745 rows x 8 columns]

In [ ]:
len(merged_df.user_id.unique())

32055

In [ ]:
timestamp_cols = ['sent_email', 'delivered_email', 'open_email', 'clicked_email', 'spam_email', 'bounced_email']
for col in timestamp_cols:
    merged_df[col] = pd.to_datetime(merged_df[col])

In [ ]:
def resample_merged_handling_nan(df, timestamp_cols, freq='M'):
    resampled_data = {}
    for col in timestamp_cols:
        temp_df = df.dropna(subset=[col])
        resampled_data[col] = temp_df.set_index(col).resample(freq).count()['user_id']
    return pd.DataFrame(resampled_data)

df_filtered = merged_df.dropna(subset=timestamp_cols, how='all')

# Resample by month, handling NaNs
df_resampled = resample_merged_handling_nan(merged_df, timestamp_cols)

# Plotting using Plotly
fig = px.line(df_resampled.reset_index(), x='index', y=timestamp_cols, title='Email Status Count Over Time (Each Month)', labels={'index': 'Time'})
fig.show()

In [ ]:
df_counts

2019-01  2019-02  2019-03  2019-04  2019-05  2019-06  \
sent_email          46.0     74.0    106.0    143.0    169.0    191.0   
delivered_email     42.0     68.0     89.0    126.0    142.0    167.0   
open_email           6.0     12.0     11.0     19.0     29.0     34.0   
clicked_email        1.0      2.0      1.0      0.0      3.0      0.0   
spam_email           4.0      3.0      1.0      4.0      8.0      8.0   
bounced_email        4.0      6.0     17.0     17.0     27.0     24.0   

                 2019-07  2019-08  2019-09  2019-10  2019-11  2019-12  \
sent_email         270.0    291.0    390.0    443.0    582.0    698.0   
delivered_email    241.0    260.0    345.0    390.0    512.0    621.0   
open_email          32.0     50.0     65.0     67.0     80.0    101.0   
clicked_email        2.0      1.0      5.0     10.0     10.0      5.0   
spam_email           9.0     10.0      6.0      9.0     28.0     18.0   
bounced_email       29.0     31.0     45.0     53.0     70.0     77.0   

                 2020-01  2020-02  2020-03  2020-04  2020-05  2020-06  \
sent_email         929.0   1019.0   1209.0   1318.0   1364.0   1343.0   
delivered_email    832.0    905.0   1081.0   1160.0   1196.0   1203.0   
open_email         127.0    165.0    168.0    187.0    198.0    236.0   
clicked_email       12.0     11.0     13.0     12.0     14.0     14.0   
spam_email          41.0     31.0     31.0     43.0     44.0     46.0   
bounced_email       97.0    114.0    128.0    158.0    168.0    140.0   

                 2020-07  2020-08  2020-09  2020-10  2020-11  2020-12  \
sent_email        1452.0   1365.0   1384.0   1350.0   1322.0   1416.0   
delivered_email   1287.0   1215.0   1218.0   1196.0   1171.0   1260.0   
open_email         235.0    231.0    193.0    209.0    205.0    201.0   
clicked_email       18.0     14.0     19.0     16.0     16.0      9.0   
spam_email          43.0     54.0     37.0     46.0     47.0     50.0   
bounced_email      165.0    150.0    166.0    154.0    151.0    156.0   

                 2021-01  2021-02  2021-03  2021-04  2021-05  2021-06  \
sent_email        1321.0   1159.0   1346.0   1266.0   1371.0   1344.0   
delivered_email   1179.0   1031.0   1184.0   1120.0   1227.0   1199.0   
open_email         186.0    165.0    188.0    176.0    189.0    204.0   
clicked_email       13.0     15.0     12.0     13.0      5.0     17.0   
spam_email          41.0     45.0     45.0     43.0     36.0     43.0   
bounced_email      142.0    128.0    162.0    146.0    144.0    145.0   

                 2021-07  2021-08  2021-09  2021-10  2021-11  2021-12  \
sent_email        1245.0   1313.0   1336.0   1303.0   1197.0   1252.0   
delivered_email   1111.0   1168.0   1184.0   1168.0   1053.0   1108.0   
open_email         192.0    191.0    190.0    199.0    175.0    186.0   
clicked_email       12.0     14.0     15.0     13.0     11.0      5.0   
spam_email          48.0     43.0     49.0     45.0     30.0     34.0   
bounced_email      134.0    145.0    152.0    135.0    144.0    144.0   

                 2022-01  2022-02  2022-03  2022-04  2022-05  2022-06  \
sent_email        1220.0   1094.0   1109.0   1036.0    928.0    813.0   
delivered_email   1079.0    992.0    979.0    920.0    825.0    714.0   
open_email         185.0    171.0    161.0    140.0    140.0    105.0   
clicked_email       15.0     11.0     16.0      8.0     10.0      5.0   
spam_email          56.0     32.0     33.0     36.0     27.0     17.0   
bounced_email      141.0    102.0    130.0    116.0    103.0     99.0   

                 2022-07  2022-08  2022-09  2022-10  2022-11  2022-12  
sent_email         701.0    615.0    457.0    288.0    149.0      8.0  
delivered_email    601.0    557.0    415.0    255.0    137.0      7.0  
open_email          99.0     87.0     76.0     51.0     25.0      5.0  
clicked_email       11.0      3.0      5.0      2.0      2.0      1.0  
spam_email          23.0     27.0     24.0     12.0      6.0      0.0  
b

In [ ]:
df_resampled_counts = pd.DataFrame()

# Resample each timestamp column individually and count occurrences, avoiding NaN values
for col in timestamp_cols:
    temp_df = df_filtered[[col]].dropna()
    temp_df[col] = temp_df[col].dt.to_period("M")
    resampled_count = temp_df[col].value_counts().sort_index()
    df_resampled_counts[col] = resampled_count

# Fill NaN values with zeros
df_resampled_counts.fillna(0, inplace=True)

# Calculate the Spam to Sent and Bounced to Sent ratios
df_resampled_counts['spam_to_sent_ratio'] = df_resampled_counts['spam_email'] / df_resampled_counts['sent_email']
df_resampled_counts['bounced_to_sent_ratio'] = df_resampled_counts['bounced_email'] / df_resampled_counts['sent_email']

# Plotting the ratios using Plotly
df_resampled_counts.index = df_resampled_counts.index.astype(str)

# Plotting the ratios using Plotly again
fig = px.line(df_resampled_counts.reset_index(), x='index', y=['spam_to_sent_ratio', 'bounced_to_sent_ratio'],
              title='Spam to Sent and Bounced to Sent Ratios Over Time (Each Month)', labels={'index': 'Time'})
fig.show()

Email Warming: The data doesn't strongly support that a successful email warming process has been conducted. A successful warming process would likely show a decreasing trend in Spam and Bounced rates, which isn't evident here.

## Question 5: How to do a Novermber 1st 2022 A/B test on September 2022

In [ ]:
merged_df.to_csv("data_for_usage.csv")